src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala

```diff
@@ -0,0 +1,423 @@
+
+    val labelEnvVars = EnvironmentHelper.labelsToEnvVars(labels)
+
+    // Variables defined on task level should override ones defined at pod level.
+    // Therefore the order here is important. Values for existing keys will be overwritten in the order they are added.
+    val envVars = (podEnvVars ++
+      taskEnvVars ++
+      hostEnvVar ++
+      taskContextEnvVars ++
+      labelEnvVars ++
+      portsEnvVars)
+      .map {
+        case (name, value) =>
+          mesos.Environment.Variable.newBuilder.setName(name).setValue(value).build
+      }
+
+    commandInfo.setEnvironment(mesos.Environment.newBuilder.addAllVariables(envVars.asJava))
+
+    commandInfo
+  }
+
+  private[this] def computeContainerInfo(
+    podVolumes: Seq[raml.Volume],
+    container: raml.MesosContainer): mesos.ContainerInfo.Builder = {
+    val containerInfo = mesos.ContainerInfo.newBuilder
+      .setType(mesos.ContainerInfo.Type.MESOS)
+
+    container.volumeMounts.foreach { volumeMount =>
+      podVolumes.find(_.name == volumeMount.name).map { hostVolume =>
+        val volume = mesos.Volume.newBuilder
+          .setContainerPath(volumeMount.mountPath)
+
+        hostVolume.host.foreach(volume.setHostPath)
+
+        // Read-write mode will be used when the "readOnly" option isn't set.
+        val mode = if (volumeMount.readOnly.getOrElse(false)) mesos.Volume.Mode.RO else mesos.Volume.Mode.RW
+
+        volume.setMode(mode)
+
+        containerInfo.addVolumes(volume)
+      }
+    }
+
+    container.image.foreach { im =>
+      val image = mesos.Image.newBuilder
+
+      im.forcePull.foreach(forcePull => image.setCached(!forcePull))
+
+      im.kind match {
+        case raml.ImageType.Docker =>
+          val docker = mesos.Image.Docker.newBuilder.setName(im.id)
+
+          image.setType(mesos.Image.Type.DOCKER).setDocker(docker)
+        case raml.ImageType.Appc =>
+          // These labels are necessary for AppC images to work.
+          // Given that Docker only works under linux with 64bit,
+          // let's (for now) set these values to reflect that.
+          val labels = mesos.Labels.newBuilder
+            .addAllLabels(
+              Iterable(
+              mesos.Label.newBuilder.setKey("os").setValue("linux").build,
+              mesos.Label.newBuilder.setKey("arch").setValue("amd64").build
+            ).asJava)
+
+          val appc = mesos.Image.Appc.newBuilder.setName(im.id).setLabels(labels)
+
+          image.setType(mesos.Image.Type.APPC).setAppc(appc)
+      }
+
+      val mesosInfo = mesos.ContainerInfo.MesosInfo.newBuilder.setImage(image)
+      containerInfo.setMesos(mesosInfo)
+    }
+
+    containerInfo
+  }
+
+  private[this] def computeHealthCheck(
+    healthCheck: raml.HealthCheck,
+    endpoints: Seq[raml.Endpoint]): mesos.HealthCheck.Builder = {
+    val builder = mesos.HealthCheck.newBuilder
+
+    if (healthCheck.command.isDefined) {
```

Comment:
would rather see healthCheck.command.foreach { ... } (and likewise for healthCheck.http and .tcp); leave it up to some other validator to enforce the "only 1 can be set rule" and get rid of the if isDefined checking here